File: 2015/day8/ruby/logic_test.rb
# frozen_string_literal: true

require 'test/unit'
require './logic'

# Tests the logic of day8 ruby implementation
class LogicTest < Test::Unit::TestCase
  def test_count_memory_chars # rubocop:disable Metrics/MethodLength
    assert_equal(
      2,
      count_memory_chars('\"\"')
    )
    assert_equal(
      2,
      count_memory_chars('\"abc\"')
    )
    assert_equal(
      3,
      count_memory_chars('\"aaa\\\"aaa\"')
    )
    assert_equal(
      5,
      count_memory_chars('\"\\x27\"')
    )
  end

  def test_count_reencode_string_chars # rubocop:disable Metrics/MethodLength
    assert_equal(
      6,
      count_reencode_string_chars('\"\"')
    )
    assert_equal(
      6,
      count_reencode_string_chars('\"abc\"')
    )
    assert_equal(
      8,
      count_reencode_string_chars('\"aaa\\"aaa\"')
    )
    assert_equal(
      7,
      count_reencode_string_chars('\"\\x27\"')
    )
  end
end

File: 2015/day8/go/logic/logic_test.go
package logic

import (
	"testing"
)

func TestCountMemoryChars(t *testing.T) {
	cases := []struct {
		in   string
		want int
	}{
		{`""`, 2},
		{`"abc"`, 2},
		{`"aaa\"aaa"`, 3},
		{`"\x27"`, 5},
	}

	for _, c := range cases {
		got := CountMemoryChars(c.in)
		if got != c.want {
			t.Errorf("CountMemoryChars(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}

func TestCountReencodeStringChars(t *testing.T) {
	cases := []struct {
		in   string
		want int
	}{
		{`""`, 4},
		{`"abc"`, 4},
		{`"aaa\"aaa"`, 6},
		{`"\x27"`, 5},
	}

	for _, c := range cases {
		got := CountReencodeStringChars(c.in)
		if got != c.want {
			t.Errorf("CountReencodeStringChars(%v) == %v, want %v", c.in, got, c.want)
		}
	}
}
